File: php/circular_linked_list.php
<?php
class Node
{
    public $data; // Value of the node
    public $next; // Pointer to the next node

    public function __construct($data)
    {
        $this->data = $data;
        $this->next = null; // Initially, the node points to nothing
    }
}

class CircularLinkedList
{
    private $head; // Points to the first node

    public function __construct()
    {
        $this->head = null; // Initially, the list is empty
    }

    public function addNode($data)
    {
        $newNode = new Node($data);

        if ($this->head === null)
        {
            // If the list is empty, the new node points to itself
            $this->head = $newNode;
            $newNode->next = $this->head;
        }
        else
        {
            $current = $this->head;

            // Traverse to the last node
            while ($current->next !== $this->head)
            {
                $current = $current->next;
            }

            // Update the links to add the new node
            $current->next = $newNode;
            $newNode->next = $this->head;
        }
    }

    public function display()
    {
        if ($this->head === null)
        {
            echo "The list is empty." . PHP_EOL;
            return;
        }

        $current = $this->head;
        do
        {
            echo $current->data . " -> ";
            $current = $current->next;
        }
        while ($current !== $this->head);

        echo "(back to head)" . PHP_EOL;
    }

    public function addAtBeginning($data)
    {
        $newNode = new Node($data);

        if ($this->head === null)
        {
            // If the list is empty, the new node points to itself
            $this->head = $newNode;
            $newNode->next = $this->head;
        }
        else
        {
            $current = $this->head;

            // Traverse to the last node
            while ($current->next !== $this->head)
            {
                $current = $current->next;
            }

            // Update the links to add the new node
            $newNode->next = $this->head;
            $current->next = $newNode;
            $this->head = $newNode;
        }
    }

    public function find($value)
    {
        if ($this->head === null)
        {
            return false; // List is empty
        }

        $current = $this->head;

        do
        {
            if ($current->data === $value)
            {
                return true; // Value found
            }
            $current = $current->next;
        }
        while ($current !== $this->head);

        return false; // Value not found
    }

    public function delete($value)
    {
        if ($this->head === null)
        {
            echo "The list is empty." . PHP_EOL;
            return;
        }

        $current = $this->head;
        $previous = null;

        do
        {
            if ($current->data === $value)
            {
                if ($previous === null)
                {
                    // Deleting the head node
                    $last = $this->head;

                    // Traverse to the last node
                    while ($last->next !== $this->head)
                    {
                        $last = $last->next;
                    }

                    // Update the last node to point to the new head
                    $this->head = $this->head->next;
                    $last->next = $this->head;
                }
                else
                {
                    // Deleting a middle or last node
                    $previous->next = $current->next;
                }

                unset($current);
                return;
            }
            $previous = $current;
            $current = $current->next;
        }
        while ($current !== $this->head);

        echo "Value not found in the list." . PHP_EOL;
    }

    public function countNodes()
    {
        if ($this->head === null)
        {
            return 0; // List is empty
        }

        $count = 0;
        $current = $this->head;

        do
        {
            $count++;
            $current = $current->next;
        }
        while ($current !== $this->head);

        return $count;
    }

    public function insertAt($data, $position)
    {
        $newNode = new Node($data);

        if ($position === 0)
        {
            $this->addAtBeginning($data);
            return;
        }

        $current = $this->head;
        $index = 0;

        while ($index < $position - 1 && $current->next !== $this->head)
        {
            $current = $current->next;
            $index++;
        }

        if ($index !== $position - 1)
        {
            echo "Position out of bounds." . PHP_EOL;
            return;
        }

        $newNode->next = $current->next;
        $current->next = $newNode;
    }
}

$list = new CircularLinkedList();

// Add nodes
$list->addNode(10);
$list->addNode(20);
$list->addNode(30);
$list->display(); // Output: 10 -> 20 -> 30 -> (back to head)

// Add at the beginning
$list->addAtBeginning(5);
$list->display(); // Output: 5 -> 10 -> 20 -> 30 -> (back to head)

// Find a node
echo $list->find(20) ? "Found 20\n" : "20 not found\n"; // Output: Found 20

// Delete a node
$list->delete(10);
$list->display(); // Output: 5 -> 20 -> 30 -> (back to head)

// Count nodes
echo "Number of nodes: " . $list->countNodes() . PHP_EOL; // Output: 3

// Insert at position
$list->insertAt(15, 1);
$list->display(); // Output: 5 -> 15 -> 20 -> 30 -> (back to head)
?>

File: kotlin/src/CircularLinkedList.kt
class CircularNode(var value: Int)
{
    var next: CircularNode? = null // Pointer to the next node
}

class CircularLinkedList
{
    private var tail: CircularNode? = null // Tail points to the last node, whose next points to the head

    fun insertAtHead(value: Int)
    {
        val newNode = CircularNode(value)
        if (tail == null)
        {
            tail = newNode
            tail!!.next = newNode // Make the list circular
        }
        else
        {
            newNode.next = tail!!.next
            tail!!.next = newNode
        }
        println("Inserted $value at the head.")
    }

    fun insertAtTail(value: Int)
    {
        val newNode = CircularNode(value)
        if (tail == null)
        {
            tail = newNode
            tail!!.next = newNode // Make the list circular
        }
        else
        {
            newNode.next = tail!!.next
            tail!!.next = newNode
            tail = newNode // Update tail to the new last node
        }
        println("Inserted $value at the tail.")
    }

    fun traverse()
    {
        if (tail == null)
        {
            println("The list is empty.")
            return
        }
        var current = tail!!.next
        print("Circular Linked List: ")
        do
        {
            print("${current!!.value} -> ")
            current = current.next
        } while (current != tail!!.next)
        println("(back to head)")
    }

    fun deleteAtPosition(position: Int)
    {
        if (tail == null)
        {
            println("The list is empty.")
            return
        }
        if (position == 1)
        {
            // Deleting the head node
            if (tail!!.next == tail)
            {
                tail = null // Only one node in the list
            }
            else
            {
                tail!!.next = tail!!.next!!.next // Point tail to the new head
            }
            println("Deleted node at position $position")
            return
        }

        var current = tail!!.next
        var prev: CircularNode? = null
        var count = 1

        do
        {
            if (count == position)
            {
                prev!!.next = current!!.next
                if (current == tail)
                {
                    tail = prev // update tail if the last node is deleted
                }
                println("Deleted node at position $position")
                return
            }
            prev = current
            current = current!!.next
            count++
        } while (current != tail!!.next)

        println("Position $position is out of bounds.")
    }

    fun search(value: Int): Boolean
    {
        if (tail == null) return false

        var current = tail!!.next
        do
        {
            if (current!!.value == value) return true
            current = current.next
        } while (current != tail!!.next)

        return false
    }

    fun countNodes(): Int
    {
        if (tail == null) return 0

        var count = 0
        var current = tail!!.next
        do
        {
            count++
            current = current!!.next
        } while (current != tail!!.next)

        return count
    }

    fun reverse()
    {
        if (tail == null || tail!!.next == tail) return // If the list is empty or has one node

        var prev: CircularNode? = null
        var current = tail!!.next
        val head = tail!!.next

        do
        {
            val next = current!!.next
            current.next = prev
            prev = current
            current = next
        } while (current != head)

        tail!!.next = prev
        tail = head
        println("Circular Linked List reversed.")
    }
}

fun main()
{
    val circularLinkedList = CircularLinkedList()

    circularLinkedList.insertAtHead(10)
    circularLinkedList.insertAtTail(20)
    circularLinkedList.insertAtTail(30)
    circularLinkedList.insertAtHead(5)
    circularLinkedList.traverse() // Output: 5 -> 10 -> 20 -> 30 -> (back to head)

    println("Search for 20: ${if (circularLinkedList.search(20)) "Found" else "Not Found"}") // Output: Found
    println("Search for 40: ${if (circularLinkedList.search(40)) "Found" else "Not Found"}") // Output: Not Found

    circularLinkedList.deleteAtPosition(2)
    circularLinkedList.traverse() // Output: 5 -> 20 -> 30 -> (back to head)

    println("Number of nodes: ${circularLinkedList.countNodes()}") // Output: 3

    circularLinkedList.reverse()
    circularLinkedList.traverse() // Output: 30 -> 20 -> 5 -> (back to head)
}